### PR TITLE
Setup GitHub Actions for build and tests

### PR DESCRIPTION
**Pull request by zeha**
Known todos/questions:
- [x] matrix build with multiple RELEASEs and multiple HOST_RELEASEs
- [x] add goss.tap to artifacts?
- [x] add .deb to artifacts?

### CODE REVIEW

**Comment by zeha** [comment]
Not sure why actions does nothing on the main repo, maybe its turned off in the repo settings?

File: grml-debootstrap [hunk 1]
```diff
@@ -1692,6 +1692,11 @@ grub_install() {
     eend $?
   fi
 
+  if [ -n "${BOOT_APPEND}" ] ; then
+    echo "Adding BOOT_APPEND configuration ['${BOOT_APPEND}'] to /etc/default/grub."
+    sed -i "/GRUB_CMDLINE_LINUX_DEFAULT/ s#\"\$# ${BOOT_APPEND}\"#" "${MNTPOINT}/etc/default/grub"
+  fi
+
```
Comment (mika) [inline grml-debootstrap:1699]: Ah, aren't we executing `grub_install()` from chroot-script in the --vm mode?

File: tests/build-vm-and-test.sh [hunk 2]
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+# SPDX-License-Identifier: GPL-2.0
```
Comment (mika) [inline tests/build-vm-and-test.sh:2]: Hm, should we use `GPL-2.0-or-later` here, given that we have `GPL-2+` in debian/copyright? (If so, same for the other files with the SPDX header, BTW nice for taking care of this! :heart:)

File: tests/build-vm-and-test.sh [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  set -x
+  [ -x ./tests/goss ] || curl -fsSL https://goss.rocks/install | GOSS_DST="$(pwd)/tests" sh
+  sudo apt-get update
+  sudo apt-get -y install qemu-system-x86 kpartx python3-pexpect python3-serial
+  # TODO: docker.io
+  exit 0
+fi
+
+if [ ! -d ./tests ]; then
+  echo "$0: Started from incorrect working directory"
```
Comment (mika) [inline tests/build-vm-and-test.sh:28]: `>&2` maybe?

File: tests/build-vm-and-test.sh [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+# SPDX-License-Identifier: GPL-2.0
+#
+# Install an already built grml-debootstrap.deb in docker and use it to
+# build a test VM image. Then run this VM image in qemu and check if it
+# boots.
+
+set -eu -o pipefail
+
+if [ -z "${1:-}" ] || [ "${1:-}" == "--help" ] || [ "${1:-}" == "help" ]; then
+  echo "Usage: $0 setup"
```
Comment (mika) [inline tests/build-vm-and-test.sh:11]: `>&2` maybe in lines 11-14?

File: tests/build-vm-and-test.sh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  echo " then: $0 run"
+  echo " then: $0 test"
+  echo "WARNING: $0 is potentially dangerous and may destroy the host system and/or any data."
+  exit 0
```
Comment (mika) [inline tests/build-vm-and-test.sh:15]: should we maybe exit 1 then?

File: tests/build-vm-and-test.sh [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+# debootstrap to use, default empty (let grml-debootstrap decide)
+DEBOOTSTRAP="${DEBOOTSTRAP:-}"
+
+if [ "$1" == "run" ]; then
+  # Debian version on which grml-debootstrap will *run*
+  HOST_RELEASE="${HOST_RELEASE:-bookworm}"
+
+  DEB_NAME=$(ls ./grml-debootstrap*.deb || true)
+  if [ -z "$DEB_NAME" ]; then
+    echo "$0: No grml-debootstrap*.deb found, aborting"
```
Comment (mika) [inline tests/build-vm-and-test.sh:48]: `>&2`?

File: tests/build-vm-and-test.sh [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    -w /code \
+    debian:"$HOST_RELEASE" \
+    bash -c './tests/docker-install-deb.sh '"$DEB_NAME"' && ./tests/docker-build-vm.sh '"$(id -u)"' '"/code/$TARGET"' '"$RELEASE"
+
+elif [ "$1" == "test" ]; then
+  # run tests from inside Debian system
+  exec ./tests/test-vm.sh "$PWD/$TARGET" "$RELEASE"
+
+else
+  echo "$0: unknown parameters, see --help"
```
Comment (mika) [inline tests/build-vm-and-test.sh:66]: `>&2`?

File: tests/docker-build-deb.sh [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+# SPDX-License-Identifier: GPL-2.0
+#
+# Build a grml-debootstrap.deb.
+# To be run inside docker, as this script assumes it can modify the running OS.
+
+set -eu -o pipefail
+set -x
+
+if [ "${1:-}" != "--autobuild" ]; then
+  echo "$0: Only intended for CI scenarios, will destroy source files and modify running OS."
```
Comment (mika) [inline tests/docker-build-deb.sh:11]: stderr?

File: tests/docker-build-deb.sh [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  exit 1
+fi
+BUILD_NUMBER="${2:-}"
+if [ -z "$BUILD_NUMBER" ]; then
+  echo "$0: missing build number in arguments"
```
Comment (mika) [inline tests/docker-build-deb.sh:16]: stderr?

File: tests/docker-build-vm.sh [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+# SPDX-License-Identifier: GPL-2.0
+#
+# Use an already installed grml-debootstrap to build a VM image, then
+# run it in qemu. Installs goss inside the VM.
+
+set -eu -o pipefail
+
+if [ "$#" -ne 3 ]; then
+  echo "$0: Invalid arguments"
+  echo "Expect: $0 HOST_UID TARGET RELEASE"
```
Comment (mika) [inline tests/docker-build-vm.sh:11]: stderr?

File: tests/docker-install-deb.sh [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+# SPDX-License-Identifier: GPL-2.0
+#
+# Install an already built grml-debootstrap.deb.
+# Wrapper around apt-get install for usage inside docker.
+
+set -eu -o pipefail
+
+if [ "$#" -ne 1 ]; then
+  echo "$0: Invalid arguments"
+  echo "Expect: $0 DEB_NAME"
```
Comment (mika) [inline tests/docker-install-deb.sh:11]: stderr?

File: tests/gha-build-deb.sh [hunk 12]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+# SPDX-License-Identifier: GPL-2.0
+#
+# Entrypoint for GitHub Actions to build a grml-debootstrap.deb.
+
+set -eu -o pipefail
+set -x
+
+if [ -z "${CI:-}" ] || [ -z "${GITHUB_RUN_NUMBER:-}" ]; then
+  echo "Running outside of CI pipeline."
```
Comment (mika) [inline tests/gha-build-deb.sh:10]: stderr?

File: tests/test-vm.sh [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+# SPDX-License-Identifier: GPL-2.0
+#
+# Run previously built VM image in qemu and check if it boots.
+# Requires virtiofs support in VM.
+
+set -eu -o pipefail
+
+if [ "$#" -ne 2 ]; then
+  echo "$0: Invalid arguments"
+  echo "Expect: $0 VM_IMAGE VM_HOSTNAME"
```
Comment (mika) [inline tests/test-vm.sh:11]: stderr?

File: .github/workflows/test-build.yml [hunk 14]
```diff
@@ -51,6 +51,7 @@ jobs:
         - trixie
         - bookworm
         - bullseye
+        - buster
```
Comment (mika) [inline .github/workflows/test-build.yml:54]: Is there any reason not to squash this into commit 14b94d049c9c7c545754c2d6b5709cd218891c55?

File: .github/workflows/test-build.yml [hunk 15]
```diff
@@ -42,6 +42,7 @@ jobs:
       fail-fast: false
       matrix:
         host_release:
+        - unstable
```
Comment (mika) [inline .github/workflows/test-build.yml:45]: Is it worth squashing this into commit 14b94d049c9c7c545754c2d6b5709cd218891c55?

**Review by mika — COMMENTED** [review]
Really only nitpicks, *awesome* work! Thanks for taking care! :heart:

File: .github/workflows/test-build.yml (hunk 14 above)
Comment (zeha) [reply .github/workflows/test-build.yml:54]: Was pondering if we want to keep it. I'll squash it.

File: grml-debootstrap (hunk 1 above)
Comment (zeha) [reply grml-debootstrap:1699]: Apparently not. I was wondering about the code duplication in chroot-script<>grml-debootstrap.

File: tests/build-vm-and-test.sh (hunk 2 above)
Comment (zeha) [reply tests/build-vm-and-test.sh:2]: Yes, done.

File: .github/workflows/test-build.yml (hunk 14 above)
Comment (zeha) [reply .github/workflows/test-build.yml:54]: squashed

File: .github/workflows/test-build.yml (hunk 15 above)
Comment (zeha) [reply .github/workflows/test-build.yml:45]: squashed

File: tests/build-vm-and-test.sh (hunk 3 above)
Comment (zeha) [reply tests/build-vm-and-test.sh]: done

File: tests/build-vm-and-test.sh (hunk 7 above)
Comment (zeha) [reply tests/build-vm-and-test.sh]: done

File: tests/docker-build-deb.sh (hunk 8 above)
Comment (zeha) [reply tests/docker-build-deb.sh]: done

File: tests/docker-install-deb.sh (hunk 11 above)
Comment (zeha) [reply tests/docker-install-deb.sh]: done

File: tests/test-vm.sh (hunk 13 above)
Comment (zeha) [reply tests/test-vm.sh]: done

File: tests/gha-build-deb.sh (hunk 12 above)
Comment (zeha) [reply tests/gha-build-deb.sh]: done

File: tests/docker-build-vm.sh (hunk 10 above)
Comment (zeha) [reply tests/docker-build-vm.sh]: done

File: tests/build-vm-and-test.sh (hunk 6 above)
Comment (zeha) [reply tests/build-vm-and-test.sh]: done

**Comment by mika** [comment]
Awesome work, merged, thanks @zeha! :heart:

**Pull request merged**